src/app/archive/lib/types.ml

```diff
@@ -0,0 +1,183 @@
+open Core
+open Coda_base
+open Signature_lib
+
+(** Library used to encode and decode types to a Hasura's graphql type.
+
+    Unfortunately, the graphql_ppx does not have an "encode" attribute that
+    allows us to map an OCaml type into some graphql schema type in a clean
+    way. Therefore, we have to make our own encode and decode functions. On top
+    of this, the generated GraphQL schema constructed by Hasura creates insert
+    methods where the input fields are optional, even though the inputs are
+    explicitly labeled as NOT NULL in Postgres. Additionally, graphql_ppx does
+    not let us redefine a new type to make these optional inputs mandatory.
+    Therefore, we are forced to lift the option types to some of our types.
+    Furthermore, some types that are in Postgres but are not GraphQL primitive
+    types are treated as custom scalars. These types include `bigint`, `bit(n)`
+    and enums. graphql_ppx treats custom scalars as Yojson.Basic.t types
+    (usually they are encoded as Json string types).
+
+    As a result, encoding a type to a Hasura input is broken into two phases:
+    1. Postgres phase: This phase converts OCaml type (such as blocks and
+    transactions) to some intermediate representation that is very similar to
+    their respective Postgres schema type. 2. Graphql object phase: The
+    intermediate Postgres types are converted directly to some input that the
+    Hasura Graphql schema would accept. This is essentially lifting types with
+    the Option type and coercing Postgres custom scalar into Yojson types
+
+    TODO: Another issue that occurs with Hasura is that primary serial integer
+    types, which acts as a counter whenever a new row is added, are
+    automatically generated as arguments for insert methods. This can be quite
+    problematic using this with `graphql_ppx` as these serial arguments are
+    required to `int option` for insert functions. Intuitively, the input
+    should be set as a None/Null value to automatically increment the serial
+    value. When Hasura sends this `None` value to Postgres, Postgres thinks the
+    insert query is adding a new row with its primary key equal to null.
+    Postgres thinks the query is violating a Non-null constraint and ultimately
+    rejects the insert query. Temporarily, the schemas will not have serial
+    primary key values. To address this problem in the future, it would help to
+    write a script that removes the primary serial integer arguments for insert
+    functions in the generated introspective json files. **)
```

Comment:
I think we wanted hasura because it was supposed to be easier than the raw sql. Given all this, should we still go with Hasura to talk to the db?

Reply:
The founder of Hasura responded to this issue and recommended a solution. It seems to work :). So this issue is not a problem anymore.

https://github.com/hasura/graphql-engine/issues/2908

Reply:
@bkase to answer your question. I still think it could be easier than raw sql.